Add pdf2svg imager

dalcde opened the pull request:
Currently this uses pdfcrop to crop the pdf before converting it to svg.
I've been unable to find a straightforward way to crop svg's; something
we should probably look into in the future (the Image.crop function
doesn't work on svg's. We should fix that).

The resulting svg's are much smaller than that of dvisvgm, but this is configurable via the scale factor

PatrickMassot commented on plasTeX/Imagers/__init__.py:
```diff
@@ -518,12 +518,14 @@ def writePreamble(self, document):
-            if not proc.close():
-                return True
-            return False
+        if self.verifications:
```
This is not directly related to this PR, but since you changed something here, maybe it would be a good opportunity to include more feed-back when this fails. For instance, instead of returning `False` you could return the verification that failed, and then in [`Renderer.__init__`](https://github.com/plastex/plastex/blob/acaa91fbaee4bd826a941d3a8fce18e9a4a18415/plasTeX/Renderers/__init__.py#L450-L452) change the test and write what failed when it fails. If you choose to do that then don't forget to update [the docs](https://github.com/plastex/plastex/blob/8ebba0fa05666da8b0928ea2a40f389f6a64e454/Doc/imager-api.tex#L179-L183).

dalcde replied:
I suppose one way to do this is for `verify` to raise an exception when it is missing something, and the `Renderer` would catch this error and print the corresponding message.

PatrickMassot replied:
Yes, let's keep that for a future PR.

PatrickMassot commented on plasTeX/Imagers/__init__.py:
```diff
@@ -636,12 +638,17 @@ def compileLatex(self, source):
-                output = open('images'+ext, 'rb')
+            try:
+                with open('images'+ext, 'rb') as f:
+                    output = open('images'+ext, 'rb').read()
```
This is a pretty risky change. Did you triple checked you propagated it everywhere?

dalcde replied:
Yeah. I grep'ed the source and made sure it is only propagated to `executeConvert`, and I modified all the `executeConvert`'s.

PatrickMassot commented on plasTeX/Imagers/__init__.py:
```diff
@@ -636,12 +638,17 @@ def compileLatex(self, source):
 
         # Change back to original working directory
         os.chdir(cwd)
+        shutil.rmtree(tempdir, True)
```
This is yet another unrelated change, right? It is often useful to keep those temporary stuff for debugging, and I don't think they are so large (and people are still free to wipe out `/tmp` from times to times).

dalcde replied:
I also use this for debugging sometimes, but it is less helpful if the `/tmp/` is full of things called `tmp....`. Personally, I use this on a shared host, and it would be irresponsible to pollute `/tmp/` with all my stuff. I'll add an option that removes the temp directories by default.

PatrickMassot replied:
Ok, good.

PatrickMassot commented on plasTeX/Imagers/pdf2svg.py:
```diff
@@ -0,0 +1,56 @@
+width_re = re.compile('width="([0-9\\.]*)')
+height_re = re.compile('height="([0-9\\.]*)')
+class PDFSVG(VectorImager):
+    """ Imager that uses dvisvgm """
```
Copy-paste error?

PatrickMassot commented on plasTeX/Imagers/pdf2svg.py:
```diff
@@ -0,0 +1,56 @@
+            scaled = False
+            with open("scaled-%s" % filename, "w") as ofile:
+                for line in open(filename, "r"):
+                    if not scaled and "<svg" in line:
```
All this SVG parsing seems brittle (see also the first answer on [this page](https://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags)). In the current TikZ handling we use BeautifulSoup. Maybe this is a bit extreme, but there is xml parsing in the python standard lib.

dalcde replied:
I agree. My original rationale was that we only had to parse very specific output, namely what pdf2svg gives us, but I suppose this doesn't excuse bad code. I'll fix this.

dalcde commented:
All done

PatrickMassot commented on plasTeX/Imagers/pdf2svg.py:
```diff
@@ -0,0 +1,46 @@
+    compiler = 'pdflatex'
+
+    def executeConverter(self, output):
+        open('images-uncropped.pdf', 'wb').write(output)
```
I know plasTeX is full of code like this. But it was written 15 years ago, let's not add more. Please use 
```python
with open('images-uncropped.pdf', 'wb') as uncropped:
    uncropped.write(output)
```
which makes sure to close the file handler immediately, or even more modern: start the file with `from pathlib import Path` and then use `Path('images-uncropped.pdf').write_bytes(output)` instead of the above line, since the next plasTeX release will support only python at least 3.5.

dalcde replied:
Glad to hear about the last part. Does that mean I'm allowed to add type hints as I go?

PatrickMassot replied:
Yes! You can add as many type hints as you wish, our lives would be much easier with type hints in plasTeX. Even without using mypy, it's already great documentation. But let's keep it python 3.5 compatible for now (so no PEP 526 variable type hints).

PatrickMassot commented on plasTeX/Imagers/pdf2svg.py:
```diff
@@ -0,0 +1,46 @@
+
+        rc = 0
+        page = 1
+        while 1:
```
I don't understand why this is a `while` loop. Isn't it iterating over `self.images`? Or at least `range(1, len(self.images)+1)` since you seem to use only the page number.

dalcde replied:
Yeah I think we should do that instead... I really shouldn't blindly copy existing code.

dalcde commented:
On Tue, Mar 31, 2020 at 03:35:28AM -0700, Patrick Massot wrote:
>It would be nice to add a test case. Please ask if you don't know how to write tests here.

I've looked into the testing but couldn't figure out how it works. Some
pointers would be helpful.

PatrickMassot commented on plasTeX/Imagers/__init__.py:
```diff
@@ -638,12 +638,18 @@ def compileLatex(self, source):
 
         # Change back to original working directory
         os.chdir(cwd)
+        if not self.config["images"]["save-file"]:
```
Now I'm confused. I was about to write this should be added in https://github.com/plastex/plastex/blob/master/plasTeX/Config.py but I see it's already there, added 14 years ago in https://github.com/plastex/plastex/commit/d9c53162a3252d0d5651f0d3fe16d4ef76dc5ab0. Did you see it used anywhere? It would be nice to dig a bit into the git history of the project to see what happened here. In any case, did you make sure this option is mentioned in the docs?

dalcde commented:
On Tue, Mar 31, 2020 at 07:53:46AM -0700, Patrick Massot wrote:
>@PatrickMassot commented on this pull request.
>
>
>
>>
>         # Change back to original working directory
>         os.chdir(cwd)
>+        if not self.config["images"]["save-file"]:
>
>Now I'm confused. I was about to write this should be added in https://github.com/plastex/plastex/blob/master/plasTeX/Config.py but I see it's already there, added 14 years ago in https://github.com/plastex/plastex/commit/d9c53162a3252d0d5651f0d3fe16d4ef76dc5ab0.

I grep'ed the whole codebase and it isn't used anywhere. The option
shows up when you run `plastex --help`, but it isn't in the docs proper.
I've just added it.

PatrickMassot commented on plasTeX/Renderers/__init__.py:
```diff
@@ -472,6 +472,8 @@ def render(self, document, postProcess=None):
+            elif name == 'pdf2svg':
+                from plasTeX.Imagers.pdf2svg import Imager as VectorImager
             else:
                 log.warning("Invalid imager '%s'")
```
Forget about this, I'll do it.

PatrickMassot commented:
I decided it would be too complicated to tell you how to write a test for this. So I'll merge, add a test, and discuss later. I did add a minimal [CONTRIBUTE.md](https://github.com/plastex/plastex/blob/master/CONTRIBUTING.md) that you can already read.

PatrickMassot commented:
Thanks again!

PatrickMassot commented:
@dalcde please see d40b7fbb7b40a577961205d036338f223e5fb4bf about testing an imager.
